# Reject malformed and truncated input decks
#
#   - allocation: reserve a temperature slot for *SPECIFIC GAS CONSTANT
#   - readinput: reject an input deck without any keyword card
#   - allocation: do not skip continuation lines past the end of a block
#   - *NODE/EL/SECTION/CONTACT PRINT: stop reading at the end of the input
#   - calinput: stop skipping to the next keyword at the end of the input
#   - surfaces: reject non-positive element numbers, skip set members above ne
#   - couplings: require a valid REF NODE on *COUPLING
#   - mpcs: reject *MPC node numbers that do not exist
#   - noelsets: reject non-positive node/element numbers in *NSET/*ELSET
#   - usermaterials: reject data lines with more entries than CONSTANTS allows
#   - arpack: stop when dsaupd/dnaupd fail instead of calling d[n,s]eupd
#   - rigidbodys: reject negative REF NODE and ROT NODE on *RIGID BODY
#   - dloads: skip an empty element set or facial surface
#   - pretensionsections: reject an empty surface
#
# Found by fuzzing ccx with malformed variants of 518 input decks under
# AddressSanitizer, and by review of the code around each crash.

diff --git a/ccx_2.23/src/allocation.f b/ccx_2.23/src/allocation.f
index f696005..9c801d7 100644
--- a/ccx_2.23/src/allocation.f
+++ b/ccx_2.23/src/allocation.f
@@ -378,6 +378,13 @@
           ntmatl=ntmatl+1
           ntmat_=max(ntmatl,ntmat_)
         enddo
+      elseif(textpart(1)(1:20).eq.'*SPECIFICGASCONSTANT') then
+!
+!       the specific gas constant is stored in shcon(3,1,*)
+!
+        ntmat_=max(1,ntmat_)
+        call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
+     &       ipoinp,inp,ipoinpc)
       elseif(textpart(1)(1:11).eq.'*CONSTRAINT') then
         do
           call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
@@ -872,7 +879,9 @@ c          endif
      &           inl,ipoinp,inp,ipoinpc)
             if((istat.lt.0).or.(key.eq.1)) exit
             ntmatl=ntmatl+1
-            iline=iline+1
+            call getnewline(inpc,textpart,istat,n,key,iline,ipol,
+     &           inl,ipoinp,inp,ipoinpc)
+            if((istat.lt.0).or.(key.eq.1)) exit
           enddo
           ntmat_=max(ntmatl,ntmat_)
         elseif(ityp.eq.21) then
@@ -881,7 +890,12 @@ c          endif
      &           inl,ipoinp,inp,ipoinpc)
             if((istat.lt.0).or.(key.eq.1)) exit
             ntmatl=ntmatl+1
-            iline=iline+2
+            call getnewline(inpc,textpart,istat,n,key,iline,ipol,
+     &           inl,ipoinp,inp,ipoinpc)
+            if((istat.lt.0).or.(key.eq.1)) exit
+            call getnewline(inpc,textpart,istat,n,key,iline,ipol,
+     &           inl,ipoinp,inp,ipoinpc)
+            if((istat.lt.0).or.(key.eq.1)) exit
           enddo
           ntmat_=max(ntmatl,ntmat_)
         endif
@@ -1508,7 +1522,9 @@ c            nmpc_=nmpc_+3
             if((istat.lt.0).or.(key.eq.1)) exit
             ntmatl=ntmatl+1
             ntmat_=max(ntmatl,ntmat_)
-            iline=iline+1
+            call getnewline(inpc,textpart,istat,n,key,iline,ipol,
+     &           inl,ipoinp,inp,ipoinpc)
+            if((istat.lt.0).or.(key.eq.1)) exit
           enddo
         endif
       elseif(textpart(1)(1:10).eq.'*HYPERFOAM') then
@@ -1542,7 +1558,9 @@ c            nmpc_=nmpc_+3
             if((istat.lt.0).or.(key.eq.1)) exit
             ntmatl=ntmatl+1
             ntmat_=max(ntmatl,ntmat_)
-            iline=iline+1
+            call getnewline(inpc,textpart,istat,n,key,iline,ipol,
+     &           inl,ipoinp,inp,ipoinpc)
+            if((istat.lt.0).or.(key.eq.1)) exit
           enddo
         endif
       elseif(textpart(1)(2:10).eq.'KINEMATIC') then
diff --git a/ccx_2.23/src/arpack.c b/ccx_2.23/src/arpack.c
index a04b14f..7ca1b40 100644
--- a/ccx_2.23/src/arpack.c
+++ b/ccx_2.23/src/arpack.c
@@ -842,8 +842,9 @@ void arpack(double *co, ITG *nk, ITG **konp, ITG **ipkonp, char **lakonp,
     if(info!=0){
       printf(" *ERROR in d[n,s]aupd: info=%" ITGFORMAT "\n",info);
       printf("       # of converged eigenvalues=%" ITGFORMAT "\n\n",iparam[4]);
-    }         
-    
+      if(info<0) FORTRAN(stop,());
+    }
+
     NNEW(select,ITG,ncv);
     
     if(nasym==1){
diff --git a/ccx_2.23/src/calinput.f b/ccx_2.23/src/calinput.f
index 1fdbfe4..536a067 100644
--- a/ccx_2.23/src/calinput.f
+++ b/ccx_2.23/src/calinput.f
@@ -266,6 +266,13 @@
       loop: do
 !     
       if(istat.lt.0) then
+        if(ier.ge.1) then
+          write(*,*) '*ERROR in calinput: at least one fatal'
+          write(*,*) '       error message while reading the'
+          write(*,*) '       input deck: CalculiX stops.'
+          write(*,*)
+          call exit(201)
+        endif
         write(*,*)
         write(*,*) 'Job finished'
         write(*,*)
@@ -1254,7 +1261,7 @@ c
         do
           call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &         ipoinp,inp,ipoinpc)
-          if(key.eq.1) exit
+          if((istat.lt.0).or.(key.eq.1)) exit
         enddo
         ier=2
       endif
diff --git a/ccx_2.23/src/calinput_rfn.f b/ccx_2.23/src/calinput_rfn.f
index 5927870..4bac0d0 100644
--- a/ccx_2.23/src/calinput_rfn.f
+++ b/ccx_2.23/src/calinput_rfn.f
@@ -99,7 +99,7 @@
           do
             call getnewline(inpc,textpart,istatrfn,n,key,iline,ipol,inl,
      &           ipoinp,inp,ipoinpc)
-            if(key.eq.1) exit
+            if((istatrfn.lt.0).or.(key.eq.1)) exit
           enddo
           ier=2
         endif
diff --git a/ccx_2.23/src/contactprints.f b/ccx_2.23/src/contactprints.f
index 87b02e4..537955e 100644
--- a/ccx_2.23/src/contactprints.f
+++ b/ccx_2.23/src/contactprints.f
@@ -172,7 +172,7 @@ c      jout=max(jout,1)
       do
          call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &        ipoinp,inp,ipoinpc)
-         if(key.eq.1) exit
+         if((istat.lt.0).or.(key.eq.1)) exit
          do ii=1,n
             if((textpart(ii)(1:4).ne.'CSTR').and.
      &         (textpart(ii)(1:4).ne.'CELS').and.
diff --git a/ccx_2.23/src/couplings.f b/ccx_2.23/src/couplings.f
index 0aeec20..7737fab 100644
--- a/ccx_2.23/src/couplings.f
+++ b/ccx_2.23/src/couplings.f
@@ -108,6 +108,7 @@
       enddo
 !     
       name(1:1)=' '
+      irefnode=0
       do i=2,n
         if(textpart(i)(1:8).eq.'REFNODE=') then
           read(textpart(i)(9:18),'(i10)',iostat=istat) irefnode
@@ -176,6 +177,15 @@
      &       "*COUPLING%",ier)
         return
       endif
+!     
+      if(irefnode.le.0) then
+        write(*,*)
+     &       '*ERROR reading *COUPLING: no valid REF NODE given'
+        write(*,*) '  '
+        call inputerror(inpc,ipoinpc,iline,
+     &       "*COUPLING%",ier)
+        return
+      endif
 !     
       if(orientation.eq.'                    ') then
         iorientation=0
diff --git a/ccx_2.23/src/dloads.f b/ccx_2.23/src/dloads.f
index 59b92ce..b9fdcc5 100644
--- a/ccx_2.23/src/dloads.f
+++ b/ccx_2.23/src/dloads.f
@@ -457,6 +457,13 @@ c          enddo
             call bodyadd(cbody,ibody,xbody,nbody,nbody_,elset,label,
      &           iamplitude,xmagnitude,p1,p2,bodyf,xbodyold,lc,idefbody)
           else
+            if(istartset(i).gt.iendset(i)) then
+              elset(ipos:ipos)=' '
+              write(*,*) '*WARNING reading *DLOAD: element set'
+              write(*,*) '         or facial surface ',elset
+              write(*,*) '         is empty'
+              cycle
+            endif
             l=ialset(istartset(i))
             if(surface) then
               write(label(2:2),'(i1)') l-10*(l/10)
diff --git a/ccx_2.23/src/elprints.f b/ccx_2.23/src/elprints.f
index ac99506..e727138 100644
--- a/ccx_2.23/src/elprints.f
+++ b/ccx_2.23/src/elprints.f
@@ -194,7 +194,7 @@ c          enddo
       do
          call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &        ipoinp,inp,ipoinpc)
-         if(key.eq.1) exit
+         if((istat.lt.0).or.(key.eq.1)) exit
          do ii=1,n
             if(textpart(ii)(1:4).eq.'PEEQ') then
                if((nmethod.eq.2).or.(nmethod.eq.3)) then
diff --git a/ccx_2.23/src/mpcs.f b/ccx_2.23/src/mpcs.f
index 9e7fef2..7433b0c 100644
--- a/ccx_2.23/src/mpcs.f
+++ b/ccx_2.23/src/mpcs.f
@@ -146,6 +146,13 @@ c     enddo
                   inode=inode-1
                   cycle
                endif
+               if((node.lt.0).or.(node.gt.nk)) then
+                  write(*,*) '*ERROR reading *MPC: node ',node
+                  write(*,*) '       does not exist'
+                  call inputerror(inpc,ipoinpc,iline,
+     &"*MPC%",ier)
+                  return
+               endif
                if(label(1:8).eq.'STRAIGHT') then
                   call straightmpc(ipompc,nodempc,coefmpc,
      &                 labmpc,nmpc,nmpc_,mpcfree,ikmpc,ilmpc,
diff --git a/ccx_2.23/src/nodeprints.f b/ccx_2.23/src/nodeprints.f
index 0fe8a17..6d66f60 100644
--- a/ccx_2.23/src/nodeprints.f
+++ b/ccx_2.23/src/nodeprints.f
@@ -194,7 +194,7 @@
       do
          call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &        ipoinp,inp,ipoinpc)
-         if(key.eq.1) exit
+         if((istat.lt.0).or.(key.eq.1)) exit
          do ii=1,n
             if((textpart(ii)(1:4).ne.'U   ').and.
      &         (textpart(ii)(1:4).ne.'NT  ').and.
diff --git a/ccx_2.23/src/noelsets.f b/ccx_2.23/src/noelsets.f
index a8babf8..6412827 100644
--- a/ccx_2.23/src/noelsets.f
+++ b/ccx_2.23/src/noelsets.f
@@ -347,6 +347,16 @@ ccc   to remove end
 !     
 !     node or element number
 !     
+              if(ialset(nalset+1).le.0) then
+                write(*,*) 
+     &               '*ERROR reading *NSET/ELSET: value ',
+     &               ialset(nalset+1)
+                write(*,*) '       in set ',
+     &               set(iset)(1:index(set(iset),' ')-2),' <= 0'
+                call inputerror(inpc,ipoinpc,iline,
+     &               "*NSET or *ELSET%",ier)
+                return
+              endif
               if(kode.eq.0) then
                 if(ialset(nalset+1).gt.nk) then
                   write(*,*) 
diff --git a/ccx_2.23/src/pretensionsections.f b/ccx_2.23/src/pretensionsections.f
index e372d06..c2e8653 100644
--- a/ccx_2.23/src/pretensionsections.f
+++ b/ccx_2.23/src/pretensionsections.f
@@ -213,6 +213,13 @@ c     enddo
      &        '*ERROR reading *PRE-TENSION SECTION: nonexistent surface'
           write(*,*) '       or surface consists of nodes'
           call inputerror(inpc,ipoinpc,iline,
+     &         "*PRE-TENSION SECTION%",ier)
+          return
+        endif
+        if(istartset(iset).gt.iendset(iset)) then
+          write(*,*) 
+     &        '*ERROR reading *PRE-TENSION SECTION: surface is empty'
+          call inputerror(inpc,ipoinpc,iline,
      &         "*PRE-TENSION SECTION%",ier)
           return
         endif
diff --git a/ccx_2.23/src/readinput.c b/ccx_2.23/src/readinput.c
index ad8f5e6..1f6056f 100644
--- a/ccx_2.23/src/readinput.c
+++ b/ccx_2.23/src/readinput.c
@@ -476,6 +476,11 @@ void readinput(char *jobnamec,char **inpcp,ITG *nline,ITG *nset,
       }
   }while(1);
 
+  if(ikey==0){
+      printf(" *ERROR in readinput: file %s\n contains no keyword cards\n",fninp);
+      exit(201);
+  }
+
   inp[3*ipoinp[2*ikey-1]-2]=*nline;
   RENEW(inpc,char,(long long)132**nline);
   
diff --git a/ccx_2.23/src/rigidbodys.f b/ccx_2.23/src/rigidbodys.f
index 20f3141..3352a7c 100644
--- a/ccx_2.23/src/rigidbodys.f
+++ b/ccx_2.23/src/rigidbodys.f
@@ -113,7 +113,7 @@
      &           "*RIGID BODY%",ier)
             return
           endif
-          if(irefnode.gt.nk) then
+          if((irefnode.lt.0).or.(irefnode.gt.nk)) then
             write(*,*) '*ERROR reading *RIGID BODY: ref node',
      &           irefnode
             write(*,*) '       has not been defined'
@@ -127,7 +127,7 @@
      &           "*RIGID BODY%",ier)
             return
           endif
-          if(irotnode.gt.nk) then
+          if((irotnode.lt.0).or.(irotnode.gt.nk)) then
             write(*,*) '*ERROR reading *RIGID BODY: rot node',
      &           irotnode
             write(*,*) '       has not been defined'
diff --git a/ccx_2.23/src/sectionprints.f b/ccx_2.23/src/sectionprints.f
index bdf74a8..4959a15 100644
--- a/ccx_2.23/src/sectionprints.f
+++ b/ccx_2.23/src/sectionprints.f
@@ -172,7 +172,7 @@ c          enddo
       do
          call getnewline(inpc,textpart,istat,n,key,iline,ipol,inl,
      &        ipoinp,inp,ipoinpc)
-         if(key.eq.1) exit
+         if((istat.lt.0).or.(key.eq.1)) exit
          loop: do ii=1,n
             if((textpart(ii)(1:4).ne.'DRAG').and.
      &         (textpart(ii)(1:4).ne.'FLUX').and.
diff --git a/ccx_2.23/src/surfaces.f b/ccx_2.23/src/surfaces.f
index 0ea8b82..546910f 100644
--- a/ccx_2.23/src/surfaces.f
+++ b/ccx_2.23/src/surfaces.f
@@ -303,6 +303,12 @@
                 if(l.gt.0) then
                   kstart=kend
                   kend=l
+                  if(l.gt.ne) then
+                    write(*,*) '*WARNING reading *SURFACE: element ',
+     &                   l
+                    write(*,*) '         in set ',set(iset),' > ne'
+                    cycle
+                  endif
                   nalset=nalset+1
                   if(nalset.gt.nalset_) then
                     write(*,*) 
@@ -331,9 +337,10 @@
                   ialset(nalset)=iside+10*l
                 else
                   kstart=kstart
-                  nalset=nalset-1
+                  if(kend.le.ne) nalset=nalset-1
                   kincrement=-ialset(j)
                   do l=kstart+kincrement,kend,kincrement
+                    if(l.gt.ne) exit
                     nalset=nalset+1
                     if(nalset.gt.nalset_) then
                       write(*,*) 
@@ -374,7 +381,13 @@
               return
             endif
           else
-            if(l.gt.ne) then
+            if(l.le.0) then
+              write(*,*) '*ERROR reading *SURFACE: element ',
+     &             l,' is not a valid element number'
+              call inputerror(inpc,ipoinpc,iline,
+     &             "*SURFACE%",ier)
+              return
+            elseif(l.gt.ne) then
               write(*,*) '*WARNING reading *SURFACE: element ',
      &             l
               write(*,*) '         in set ',set(iset),' > ne'
diff --git a/ccx_2.23/src/usermaterials.f b/ccx_2.23/src/usermaterials.f
index b7d27af..6e943a1 100644
--- a/ccx_2.23/src/usermaterials.f
+++ b/ccx_2.23/src/usermaterials.f
@@ -122,6 +122,14 @@
                if(8*j.gt.nconstants+1) then
                   imax=nconstants-8*j+9
                endif
+               if(n.gt.imax) then
+                  write(*,*) 
+     &           '*ERROR reading *USER MATERIAL: more entries than'
+                  write(*,*) '  allowed by the CONSTANTS parameter'
+                  call inputerror(inpc,ipoinpc,iline,
+     &                 "*USER MATERIAL%",ier)
+                  return
+               endif
                do i=1,imax
                   if(isum+i.le.nconstants) then
                      read(textpart(i)(1:20),'(f20.0)',iostat=istat) 
